crypto/threads_pthread.c

```diff
@@ -42,12 +42,401 @@
+
+    /*
+     * get the current qp index
+     */
+try_again:
+    qp_idx = __atomic_load_n(&ilock->reader_idx, __ATOMIC_SEQ_CST);
+    __atomic_add_fetch(&ilock->qp_group[qp_idx].users, VAL_READER,
+                       __ATOMIC_SEQ_CST);
+    if (qp_idx != __atomic_load_n(&ilock->reader_idx, __ATOMIC_SEQ_CST)) {
+        __atomic_sub_fetch(&ilock->qp_group[qp_idx].users, VAL_READER,
+                       __ATOMIC_SEQ_CST);
+        goto try_again;
+    }
+
+    return (struct rcu_qp *)&ilock->qp_group[qp_idx];
+}
+
+/*
+ * Public READ lock api
+ */
+void CRYPTO_THREAD_rcu_read_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_thr_data *data;
+
+    /*
+     * we're going to access current_qp here so ask the
+     * processor to fetch it
+     */
+    data = pthread_getspecific(rcu_thr_key);
+
+    if (data == NULL) {
+        data = OPENSSL_zalloc(sizeof(struct rcu_thr_data));
+        assert(data != NULL);
+        pthread_setspecific(rcu_thr_key, data);
+    }
+
+    if (data->qp == NULL)
+        data->qp = get_hold_current_qp(lock);
+
+    /* inc our local count */
+    data->count++;
+
+    /* check for underflow condition */
+    assert(data->count >= 0);
+}
+
+/*
+ * Public READ unlock api
+ */
+void CRYPTO_THREAD_rcu_read_unlock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_thr_data *data = pthread_getspecific(rcu_thr_key);
+
+    assert(data != NULL);
+
+    data->count--;
+
+    assert(data->count >= 0);
+
+    if (data->count == 0) {
+        __atomic_sub_fetch(&data->qp->users, VAL_READER,
+                           __ATOMIC_SEQ_CST);
+        data->qp = NULL;
+    }
+}
+
+/*
+ * Write side allocation routine to get the current qp
+ * and replace it with a new one
+ */
+static struct rcu_qp *update_qp(struct rcu_lock_internal *ilock)
+{
+    uint64_t new_id;
+    uint32_t current_idx;
+
+    pthread_mutex_lock(&ilock->alloc_lock);
+
+    /*
+     * we need at least one qp to be available with one
+     * left over, so that readers can start working on
+     * one that isn't yet being waited on
+     */
+    while (ilock->group_count - ilock->writers_alloced < 2) {
+        /*
+         * we have to wait for one to be free
+         */
+        pthread_cond_wait(&ilock->alloc_signal, &ilock->alloc_lock);
+    }
+
+    current_idx = ilock->current_alloc_idx;
+    /*
+     * Allocate the qp
+     */
+    ilock->writers_alloced++;
+
+    /*
+     * increment the allocation index
+     */
+    ilock->current_alloc_idx =
+        (ilock->current_alloc_idx + 1) % ilock->group_count;
+
+    /*
+     * get and insert a new id
+     */
+    new_id = __atomic_fetch_add(&ilock->id_ctr, 1, __ATOMIC_SEQ_CST);
+
+    new_id = VAL_ID(new_id);
+    __atomic_and_fetch(&ilock->qp_group[current_idx].users, ID_MASK,
+                       __ATOMIC_SEQ_CST);
+    __atomic_or_fetch(&ilock->qp_group[current_idx].users, new_id,
+                      __ATOMIC_SEQ_CST);
+
+    /*
+     * update the reader index to be the prior qp
+     */
```

Comment:
The prior qp? I don't understand this, since we seem to be using "current_alloc_idx" which we just incremented?